### DISCUSSION TITLE
Evaluate using Profile-Guided Optimization (PGO)

### DISCUSSION THREAD

**Discussion by zamazan4ik** [optiprism-io/optiprism#292] [Ideas]
Hi!

Several days ago I [found](https://www.reddit.com/r/rust/comments/1fv3p5k/optiprism_free_rust_opensource_product_and/) the Reddit post about Optiprism and its technological stack.

I think applying PGO to Optiprism can help improve its performance and TCO (Total Cost of Ownership) in real production environments (as it already [help](https://github.com/zamazan4ik/awesome-pgo) other projects). According to the Reddit post above, there are already benchmarks in the `awesome-pgo` repo about optimizing various databases (including RocksDB) and similar software before and after PGO.

I suggest at first simply test the PGO approach with Optiprism. If it shows improvements - at least add a note in the documentation about optimizing Optiprism with PGO. Lately, we can try to optimize prebuilt binaries with PGO.

Thank you

**Comment by ravlio** [comment]
Great topic, thanks. I hope I'll find time to experiment with such optimizations. For now, there are a lot of things that can be improved just by fixing the code and it can gain x2-x10 performance :)

**Reply by zamazan4ik** [reply]
> For now, there are a lot of things that can be improved just by fixing the code and it can gain x2-x10 performance :)

A nice thing about PGO is that it's still helpful even after "fixing the code" in the most bottlenecked places since PGO allows to the compiler perform better low-level code generation. Humans do high-level clever stuff, the compiler does the dirty low-level things.
